### PR TITLE
be nice with scheduler when naptime = 0

### PR DESCRIPTION
**Pull request by Sasasu**
before: 

after:

with diskauota paused:

this PR avoids a busy loop in diskquota launcher when naptime = 0. save 100% (1core) CPU resource.

### CODE REVIEW

**Comment by higuoxing** [comment]
Nice catch!

File: diskquota.c [hunk 1]
```diff
@@ -359,7 +359,8 @@ disk_quota_worker_main(Datum main_arg)
 		 */
 		rc = WaitLatch(&MyProc->procLatch,
 					   WL_LATCH_SET | WL_TIMEOUT | WL_POSTMASTER_DEATH,
-					   diskquota_naptime * 1000L);
+					   // be nice to scheduler when naptime == 0 and diskquota_is_paused() == true
+					   diskquota_naptime == 0 ? 1:  diskquota_naptime * 1000L);
```
Comment (beeender) [inline diskquota.c:363]: I didn't get this. Why this change is related with `diskquota_is_paused()`?

Comment (Sasasu) [reply diskquota.c:363]: https://github.com/greenplum-db/diskquota/pull/120/files#diff-c07e8cb1219541bb7fd59c7049895dd1869e2df1c96073c8969d9d641977109bR378-R379

Comment (beeender) [reply diskquota.c:363]: Uh ... sorry,  I thought the **1** in `diskquota_naptime == 0 ? 1:  diskquota_naptime * 1000L);` is 1 second. 
1 ms is good :+1:

Comment (beeender) [reply diskquota.c:363]: Do we need the same change for the above wait?
```
rc = WaitLatch(&MyProc->procLatch,
					   WL_LATCH_SET | WL_TIMEOUT | WL_POSTMASTER_DEATH,
diskquota_naptime * 1000L);
```

**Comment by beeender** [comment]
This PR is needed. I doubt some unexpected long run test is caused by a very high CPU usage. e.g.: https://extensions.ci.gpdb.pivotal.io/teams/main/pipelines/PL:diskquota-gpdb-6X-stable/jobs/diskquota_ubuntu18_build_test/builds/14

**Comment by Sasasu** [comment]
usleep(1) still using many CPU resources, but better than before, from 100% to ~= 50%

Screenshot_20220221_114333
Screenshot_20220221_114351

most CPU usage is in accessing the process linked list in the Linux kernel, there are no effects on other processes.
my firefox is happy.

File: diskquota.c [hunk 2]
```diff
@@ -316,7 +316,8 @@ disk_quota_worker_main(Datum main_arg)
 		}
 		rc = WaitLatch(&MyProc->procLatch,
 					   WL_LATCH_SET | WL_TIMEOUT | WL_POSTMASTER_DEATH,
-					   diskquota_naptime * 1000L);
+					   // be nice to scheduler when naptime == 0 and diskquota_is_paused() == true
+					   diskquota_naptime == 0 ? usleep(1), 0 : diskquota_naptime * 1000L);
```
Comment (xuebinsu) [inline diskquota.c:320]: Combining a comma operator with a conditional operator is a little bit confusing.

**Review by xuebinsu — DISMISSED** [review]
LGTM overall. The result is similar to what I expect.

File: diskquota.c (hunk 2 above)
Comment (beeender) [inline diskquota.c:320]: I don't think this is correct?
```
int main() 
{
    int a = 0;
    a = 1, 2;
    std::cout << a;
    return 0;
}
```

```
❯ g++ main.cpp && ./a.out
1
```

Considering `usleep()` may return `-1`, we risk setting a timeout with `-1`.

just use some NON-magic code

```
WaitLatch(&MyProc->procLatch,
					   WL_LATCH_SET | WL_TIMEOUT | WL_POSTMASTER_DEATH,
					   // be nice to scheduler when naptime == 0 and diskquota_is_paused() == true
					   diskquota_naptime *= 1000L);
if (!diskquota_naptime) {
    usleep(1);
}
```

Comment (xuebinsu) [reply diskquota.c:320]: Hi @beeender I think this is correct. 

It is all about the operator precedence: `a = 1, 2` is equivalent to `(a = 1), 2`. This is why I consider it "confusing".

I agree with your suggestion.

**Comment by Sasasu** [comment]
merge ready

**Review by beeender — APPROVED** [review]

**Review by xuebinsu — APPROVED** [review]
